### ISSUE TITLE
Documentation for using e2guardian with e2guardian.pl and for HTTPS filtering

### ISSUE THREAD

**Issue by rmdtech** [e2guardian/e2guardian#76]
Judging by a search on the Internet, many users, myself included are having difficulty setting up and configuring e2guardian.pl/dansguardian.pl as most of us are used to using template.html.  However with the rise of the need to filter https traffic, using the template.html is no longer possible due to the template file not supporting https.  Thus people are trying to setup e2guardian.pl but we are all running into the same problem - very little documentation on how to set this up and get it working properly.

My current issue is that when I changed reportinglevel to 2 and setup accessdenied to point to the location of the e2guardian.pl file, I just get a browser error when connecting to a blocked website even one that is using http - never mind https.  The template.html shows access denied.  I have no idea what I need to change in the e2guardian or e2guardianf1.conf file to fix the error for http websites never mind https sites.

Would it be possible to write an updated step by step guide for setting up e2guardian.pl so that it works for http websites and works for https websites as too many people seem to be struggling with this issue - especially getting https websites to show the access denied page instead of proxy server refusing connections.

such as ./configure --"exactly what options are required here to get e2guardian.pl to work for http websites"
/configure --"exactly what options are required here to get e2guardian.pl to work for https websites"

e2guardian.conf file - exactly what changes need to be made in the default config file to get above to work.
e2guardianf1.conf - exactly what changes need to be made in the default config file to get above to work.

I have been looking around on the Internet on various websites with people asking the same question re e2guardian/dansguardian about how to get ssl filtering to work so that access denied page shows access denied etc when visiting a blocked https website.  A lot of people are saying that there is no clear step by step guide to how to correctly configure and set this up.  All too often the help is troubleshooting that specific person's setup rather than than a proper ABC guide to setting this up correctly.

I wil have to wait until someone has time to write a step by step guide to getting e2guardian.pl and ssl filtering to work as I am unable to work out what I need to compile, edit in conf file to get e2guardian.pl to display access denied page for both http and https websites.

I have setup e2guardian and proxy authentication and that is working brilliantly, but I am having to use template.html as I cannot get e2guardian.pl to work as I don't know what I am doing wrong and I am not the only one that is struggling with this.  The main problem as I see it, is that there is just not enough documentation on this issue.  Thank you

**Comment by rmdtech** [comment]
I am also wondering if maybe the problem I am having is that I have not configured Squid to support HTTPS traffic a perhaps the Squid proxy handles decrypting the traffic and passes the result on to e2Guardian to filter which then hands it back to Squid to send on to the client computer?

**Comment by rmdtech** [comment]
I think that I am now understanding how https filtering works.  I think that there are two different issue here.

Lets say I block facebook.com.  This is a banned site and the proxy server sends the domain name to the filtering server program which then returns the response "cannot connect to site using HTTPS" - because the browser is expecting an HTTPS response and is instead getting the HTTP access denied page.

We get this error message!  This is what I was trying to fix, but according to research its not possible to fix it?

The second issue is getting the proxy to be able to scan an HTTPS page i.e. a website that is not in the banned list.  I think that the issue of getting certificates and setting up squid --with-openssl is to address this issue ie to scan content on HTTPS pages.  If I have understood it correctly, we should see the proper access denied page, because the proxy server is decrypting the contents and then sending it to the filtering server which then sends the access denied page back to the proxy server and the proxy server then encrypts it and sends it back to the client who then sees the access denied page.

I think this is why I am confused, because there are two different issues here

**Comment by fredbcode** [comment]
*configure --"exactly what options are required here to get e2guardian.pl to work for https websites* -> Nothing at all

Without SSLMITM there are two different ways:

https://github.com/e2guardian/e2guardian/blob/master/configs/e2guardianf1.conf.in#L317 

Put a HTTP web page somewhere.

sslaccessdeniedaddress = 'http://YOURSERVER.YOURDOMAIN/denyssl.htm'
ssldeniedrewrite = off

The problem is that **works only with FF**

Or for any browser:

ssldeniedrewrite = on 

In this case put a http**s** web page somewhere 
sslaccessdeniedaddress = 'http**s**://YOURSERVER.YOURDOMAIN/denyssl.htm'

Without reporting levels 3 we can't do that because this is a SSL protocol restriction. 
Why ? It's very simple to understand, with reportinglevel 2 E2 switches the real page by another; how I can do that with a protected website ? You simply can't ...

With reporting level 3 E2 tells to the user "go to another website" for HTTP and same thing for HTTPS if you have sslaccessdeniedaddress configured, in this case the web page is not changed but your browser connection is redirected.

SSLIMTM is a very different thing, SSL proxy that acts as a "man in the middle". In other words, proxied HTTPS requests are terminated by the proxy and resent to the remote user, so like this we can filtering the contents, try av scanning, use the default banned page, etc

FI: https://groups.google.com/forum/#!topic/e2guardian/M2LhXerJ568

**Comment by rmdtech** [comment]
Thank you for your reply.  I am busy updating my documentation which I am going to have as a complete howto guide.  It looks like my issue may be that I was using http page instead of https page.  I will test this over the weekend when I have time.  I was thinking that maybe the problem was that I needed to edit the denyssl.htm file.

I will have another go this weekend.  Thanks

**Comment by rmdtech** [comment]
I must have made a typo in the e2guardianf1.conf file.

I have now put in

accessdeniedaddress = 'https://domain-name/cgi-bin/e2guardian.pl'
sslaccessdeniedaddress = 'https://domain-name/cgi-bin/e2guardian.pl'
ssldeniedrewrite = 'on'
reportinglevel = 3

Thanks for the tip about reportinglevel 3 (This is required for HTTPS, whereas you have to set it as 2 for HTTP).

and I am now getting the certificate is not valid message.  For something like Facebook the message I am getting is as shown in the picture
![e2guardian](https://cloud.githubusercontent.com/assets/6070032/8635224/f1d5a180-2811-11e5-9ffc-2edd76ecac35.png)

If I use reportinglevel = 2, then Http websites use the accessdeniedaddress page, but HTPPS websites come with proxy server refusing requests or whatever.  If I use Reportinglevel = 3, then HTTPS websites seem to be trying to use the access denied page, but HTTP websites are using template.html.  Is there a setting to get both HTTP and HTTPS to use the access deniedaddresspage.

If I use accessdeniedaddress = 'https:// instead of accessdeniedaddress = 'http:// Then HTTP Websites are not showing the accessdenied page, but they will show it if accessdeniedaddress = 'http://

I currently don't have an ssl certificate setup and I am hoping that by creating a certificate I might be able to fix the issue as I suspect that the problem is because my website and thus the page for the accessdeniedaddress is on a shared webserver which means that the name is not the name of my domain but rather the name of the webserver registered to that Ip address.  I hope that makes sense.

In short, is there a e2guardian.pl hosted on the internet somewhere that I can point my accessdeniedaddress to (one that is already setup with ssl certificate), to see if I have got everything with e2guardian working assuming that the only issue I actually have is that my website is not using https as I do not have an SSL Certificate.

**Comment by fredbcode** [comment]
Le 11/07/2015 22:24, rmdtech a écrit :
>
> I must have made a typo in the e2guardianf1.conf file.
>
> I have now put in
>
> accessdeniedaddress = 'https://domain-name/cgi-bin/e2guardian.pl'
> sslaccessdeniedaddress = 'https://domain-name/cgi-bin/e2guardian.pl'
> ssldeniedrewrite = 'on'
> reportinglevel = 3
>
> Thanks for the tip about reportinglevel 3 (This is required for HTTPS, 
> whereas you have to set it as 2 for HTTP).
>
> and I am now getting the certificate is not valid message. For 
> something like Facebook the message I am getting is as shown in the 
> picture
> e2guardian 
> <https://cloud.githubusercontent.com/assets/6070032/8635224/f1d5a180-2811-11e5-9ffc-2edd76ecac35.png>
>
>


https://domain-name/cgi-bin/e2guardian.pl should be a web site with a 
simple web page
Can you access https://domain-name/cgi-bin/e2guardian.pl in your browser 
without any issue ?

The picture shows something weird

Try this step by step:

1 - Just put an simple html page like this 
https://domain-name/test.html-> It's for testing purpose "hello world" 
is enough.
2 - Open this page in your browser ok ? no ? -> If no there is a problem 
somewhere, if ok step 3
3 - put in sslaccessdeniedaddress = 'https://domain-name/test.html'
4 - Restart e2guardian
5 - Try a forbidden site

After a ssl warning you should see the test.html page

Are you using transparent proxy ?

**Comment by fredbcode** [comment]
Also, facebook seems denied but what about the sub domains ? Your 
picture is provided by facebook so why you can pass ?

**Comment by rmdtech** [comment]
Sorry I think I have put too much information, so its confusing to follow.  Here is a simplified summary.

http://domain-name/cgi-bin/e2guardian.pl - this works for reportinglevel 2 and http websites

reporting level 3 uses https://domain-name/cgi-bin/e2guardian.pl for https sites, but uses template.html for http websites.  How can I get both http and https websites to use e2guardian.pl

https://domain-name/cgi-bin/e2guardian.pl -  this is showing picture number 2 for reportinglevel 2 and http websites.

https://domain-name/cgi-bin/e2guardian.pl -  for reportinglevel 2 and https websites I get a connection refused type message. Which is expected as I need reportinglevel 3 for https.

https://domain-name/cgi-bin/e2guardian.pl - this is showing either picture number one or picture number 2 for reportinglevel 3 and https websites.  For example https://en-gb.facebook.com/ shows picture one, but https://en-gb.facebook.com/login/ shows picture two.

I have tried both transparent and explicit proxy.  Transparent mode does not block facebook - you have to use explicit proxy.

I have done your test.  I created a page called e2guardiantest.html and put hello world in.  When I type www.dalacor.co.uk/e2guardiantest.html - I get hello world.  When I put https://www.dalacor.co.uk/e2guardiantest.html I get a similar picture to number 2.  so I must conclude that the issue is that I need an https certificate because the website name and host server names are different.  Also hello world does not work in the cgi-bin folder - so I moved the file to the main directory.  I suspect html files do not run in cgi-bin directory.

do you have a hosted e2guardian.pl file anywhere on the Internet that people can use to test their settings.  I am assuming that the problem is that I need to get an ssl certificate for my website, but would like to test on a working hosted site before going down the route of getting certificates.

Thank you.

**Comment by fredbcode** [comment]
Ok thanks for your explanation I pushed a fix to develop 
Keep in mind this is only a trick with several limitations, the best way is using SSLMITM

Now, I'm using reportinglevel = 3 and ssldeniedrewrte = off and sslacessdeniedadress = **http://**
The message appears only in Firefox but it's better than a blank page ...

I also tried with reportinglevel = 2, ssldeniedrewrte = on and sslacessdeniedadress = **https://** it works but don't forget that the deny webpage must be accessed through E2.

FI: About blank page, this is not an limitation from E2, but a security from SSL

**Comment by rmdtech** [comment]
I am not sure what blank page you are referring to.  I don't get blank pages.  I only get the pictures as posted above.

I have decided to leave SSLMITM for the moment as I thought that I would concentrate on fixing the issue where I could get a proper access denied page for https websites that are blocked such as facebook.  Once I have got that working, then I thought that I would move on to scanning https websites which requires SSLMITM.

I have ordered an SSL Certificate for my website as I need it for the helpdesk anyway, so I thought that I would get it now and see if it fixes the issue with the pictures above as its not displaying the page it should be displaying when using https irrespective of e2guardian.

Will keep you updated on my progress.  Thank you.

**Comment by fredbcode** [comment]
|I am not sure what blank page you are referring to. I don't get blank pages. I only get the pictures as posted above. 

I mean without any configuration, banned ssl = blank page

| I have decided to leave SSLMITM for the moment as I thought that I would concentrate on fixing the issue where I could get a proper access denied page for https websites that are blocked such as facebook. Once I have got that working, then I thought that I would move on to scanning https websites which requires SSLMITM. 

I understand, but this is not really related with SSLMITM the sslacce.. configuration is not needed.

**Comment by rmdtech** [comment]
You have lost me.  What is the sslacce..config for if it's not for showing access denied for https websites?

For blank pages - do you mean the message proxy server is refusing connections when trying to access an https website that is blocked?  I am not aware of any blank page without any configuration other than this.

**Comment by fredbcode** [comment]
Le 13/07/2015 16:34, rmdtech a écrit :
>
> You have lost me. What is the sslacce..config for if it's not for 
> showing access denied for https websites?
>

accessdeniedaddress = 'https://domain-name/cgi-bin/e2guardian.pl'
sslaccessdeniedaddress = 'https://domain-name/cgi-bin/e2guardian.pl'
ssldeniedrewrite = 'on'

Are no needed with SSLMITM activated

> For blank pages - do you mean the message proxy server is refusing 
> connections when trying to access an https website that is blocked? I 
> am not aware of any blank page without any configuration other than this.
>

When E2guardian blocks a web page the result page it changed by another, 
with SSL it can't do that so the result is a blank page, I mean no 
result at all.
Remove the three lines above and try a denied web site, the result is a 
page without any message.

Sorry, I'm probably not being very clear  ...

**Comment by rmdtech** [comment]
Ok, I am with you now.  However, maybe because I am using Firefox.  When I access a denied website with the above settings off, I get the message that proxy server is refusing connection.  I do not get a blank page.  Whether I am getting a blank page or proxy server refusing is not really important because I intend to use the accessdenied settings to get the proxy server to show the access denied page.  So I am not worried about that.

I understand you to mean that I can use the template.html file if I have SSLMITM activated?  Is that correct.  I am not ready to implement SSLMITM as this would require certificates for all proxy servers and importing the certificates into all the browsers.  At the moment, I am not ready to go down that route.

All I want to do at this point in time is block https websites like facebook.com and get a proper access denied message.  I don't currently need to scan https websites.  Also I would like Phil to complete his bug fixes as his release is a pre-release not a stable version.

**Comment by rmdtech** [comment]
Implementing the SSL Certificate and making my website HTTPS seems to have fixed most of the problems.  Now with ssldeniedrewrite = 'off' and reportinglevel =3 when I go to facebook, it comes up with the access denied page.  It is clear that you have to have a certificate and https working on your website (I presume that this is because my website was on a shared ip address - its now a dedicated ip address).

However, having ssldeniedrewrite = 'off' means that it only works in Firefox.  When I set it to on, it does not work in any browser including firefox!  I will keep working on the problem.  But in short, my solution was to make my website support https and install the SSL certificate then it works as long as ssldeniedwrite is to off.  The two pictures were because I did not have https and SSL certificate on website.

**Comment by fredbcode** [comment]
Please, if you can, try the latest dev version

**Comment by rmdtech** [comment]
Version 3.1.2 on 17th dec 2014?  I am using that version.

**Comment by fredbcode** [comment]
No this https://github.com/e2guardian/e2guardian/tree/develop branch:develop the next stable release 
I can put a tar.gz somewhere if needed

**Comment by rmdtech** [comment]
I have donwloaded the winzip file.  I will pop into a usb stick and install it on my test machine.  can I just backup an restore the lists directory and the e2guardian.conf and e2guardianf1.conf and e2guardianf2.conf files or will I need to edit the new e2guardian.conf files and lists files?  Thank you.

**Comment by fredbcode** [comment]
No, this is the stable release, dev branch can be only accessed with git 
commands

I will make a zip for you tomorrow

**Comment by rmdtech** [comment]
Oh I did not realise that I had to do something with git commands to get the dev branch.  Yes please if you could make either a tarball or zip file (whichever is quicker) that would be great.

**Comment by fredbcode** [comment]
http://numsys.eu/TMP/e2guardian-07-2015.tar.gz

**Comment by rmdtech** [comment]
Hi Fred, I downloaded your version, but with the same result.  I have decided that it would be best if I use ssldeniedrewrite = 'off' for the time being on the grounds that all my clients use Firefox - so this will work and b: I agree that ultimately implementing MITM is the best solution long term.  However I am aware that there are issues with MITM (not just with e2guardian - other companies have mentioned similar issues), so I do not plan on implementing MITM this year.

As Firefox works (which my clients use) and IE is blocking the blocked website anyway (albeit without the proper error message), I will leave it at this stage until I implement MITM which may be next year.

Thank you for all your help.  It is appreciated.  I will be writing up a complete document HOWTO which I will post on my website and link to here somewhere as there are a lot of things that I have picked up over the last few years that other people will find useful.
